provider/gce/google/raw.go

```diff
@@ -152,19 +161,54 @@ func (rc *rawConn) UpdateFirewall(projectID, name string, firewall *compute.Fire
 	if err != nil {
 		return errors.Trace(err)
 	}
+	if err := logOperationErrors(operation); err != nil {
+		return err
+	}
 
-	err = rc.waitOperation(projectID, operation, attemptsLong)
+	err = rc.waitOperation(projectID, operation, attemptsLong, logOperationErrors)
 	return errors.Trace(err)
 }
 
+type handleOperationErrors func(operation *compute.Operation) error
+
+func returnNotFoundOperationErrors(operation *compute.Operation) error {
+	hadOtherErrors := false
+	if operation.Error != nil {
+		for _, err := range operation.Error.Errors {
+			if err.Code == "RESOURCE_NOT_FOUND" {
+				return errors.NotFoundf("resource", err.Message)
+			}
+			hadOtherErrors = true
+			logger.Errorf("GCE operation error: (%s) %s", err.Code, err.Message)
+		}
+	}
+	if hadOtherErrors {
+		return waitError{operation, nil}
+	}
+	return nil
+}
+
+func logOperationErrors(operation *compute.Operation) error {
+	if operation.Error != nil {
+		for _, err := range operation.Error.Errors {
+			logger.Errorf("GCE operation error: (%s) %s", err.Code, err.Message)
+		}
+		return waitError{operation, nil}
+	}
+	return nil
+}
+
 func (rc *rawConn) RemoveFirewall(projectID, name string) error {
 	call := rc.Firewalls.Delete(projectID, name)
 	operation, err := call.Do()
 	if err != nil {
 		return errors.Trace(convertRawAPIError(err))
 	}
+	if err := returnNotFoundOperationErrors(operation); err != nil {
```

Comment:
I don't understand why you need to call this here *and* pass it into waitOperation - why isn't just calling it from waitOperation enough? (Same for the other places with this pattern.)

Reply:
Because both call.Do() and rc.waitOperation()... run operation... The first run actually happens before the wait call... Essentially, all the wait call does is to check on operation progress on cloud side. 
Since there is a chance that we'd have errors on operation already, we want to process them both after the .Do() cll and in wait...

Reply:
Actually, looking at the code, we want to give several chances to the operation to succeed. Consequently, we ignore errors until all attempts have been exhausted.

I'll change the code to only inspect once as before.